pyqtgraph/colorRegistry.py

```diff
@@ -0,0 +1,354 @@
+        # print('updating color to match',desc)
+        name, alpha = desc
+        if name[0] != '#':
+            qcol.setRgba( self.color_dic[name].rgba() )
+        else:
+            qcol.setNamedColor(name) # set from hex string
+        if alpha is not None: qcol.setAlpha(alpha)
+        # print('= hex:', qcol.name() )
+
+    def _update_QPen(self, qpen, desc):
+        """ updates qpen to match descriptor for current palette """
+        # print('updating pen to match',desc)
+        name, width, alpha = desc
+        if name[0] != '#':
+            if alpha is None:
+                qpen.setColor( self.color_dic[name] ) # automatically copies
+            else:
+                qcol = QtGui.QColor(self.color_dic[name]) # make a copy
+                qcol.setAlpha(alpha)
+                qpen.setColor(qcol)
+        else: # set from hex string instead:
+            qcol = QtGui.QColor(name)
+            if alpha is not None: qcol.setAlpha(alpha)
+            qpen.setColor(qcol)
+        if width is not None:
+            qpen.setWidth(width)
+        # print('= hex:', qpen.color().name() )
+
+    def _update_QBrush(self, qbrush, desc):
+        """ updates qbrush to match descriptor for current palette """
+        # print('updating brush',qbrush,'to match',desc)
+        name, alpha = desc
+        if name[0] != '#':
+            if alpha is None:
+                qbrush.setColor( self.color_dic[name] ) # automatically copies
+            else:
+                qcol = QtGui.QColor(self.color_dic[name]) # make a copy
+                qcol.setAlpha(alpha)
+                qbrush.setColor(qcol)
+        else: # set from hex string instead:
+            qcol = QtGui.QColor(name)
+            if alpha is not None: qcol.setAlpha(alpha)
+            qbrush.setColor(qcol)
+        # print('= hex:', qbrush.color().name(), qbrush.color().alpha() )
+
+    def register(self, obj, desc):
+        """
+        Registers obj (QColor, QPen or QBrush) to be updated according to the descriptor on palette change
+        """
+        if hasattr(obj,'registration'):
+            registration = obj.registration
+        else:
+            registration = next(ColorRegistry._registrationGenerator)
+            obj.registration = registration # patch in attribute
+        fin = weakref.finalize(obj, self.unregister, registration)
+        fin.atexit = False # no need to clean up registry on program exit
+        # print('registering', registration, '(',str(obj),'):',str(desc))
+        self.registered_objects[registration] = (weakref.ref(obj), desc)
+
+    def unregister(self, registration):
+        """
+        Removes obj (QColor, QPen or QBrush) from the registry, usually called by finalize on deletion
+        """
+        obj, desc = self.registered_objects[registration]
+        # print('unregistering', registration, '(',str(obj),'):',str(desc))
+        del self.registered_objects[registration]
+        del obj, desc
```

Comment:
Do we delete these here due to pyside bindings holding the extra reference preventing the garbage collector from grabbing them?

Reply:
No, this is just paranoia. `obj` and `desc` should get marked for garbage collection and purged from the registry by the `finalize()` code as soon as the `unregister()` method is done.

We can simply avoid creating the extra references here by commenting out the first line (#314), too.
Although I would like to keep the code as comments, since I am pretty sure we'll need it to hunt down some issue with the weakref management in the future.

The issue with PySide shows up in the testing code:
If I create and register a number of color in PyQt5, then `finalize()` calls completely clear the registry once I delete them.
In PySide, the registry still holds a single living weakref to the latest registered color. My expectation is that this will be successfully removed as soon as a new QColor is defined and/or registered. So, for now, the tests accept a <single> surviving reference. We'll need to keep an eye on this, but for now it seems to be reasonable.